torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -290,13 +290,32 @@ auto handle_torch_function_no_python_arg_parser(
   py::object ret;
   py::object mode_obj;
 
+  // Of our overloaded types, filter out the FakeTensor arguments.
+  // FakeTensor is special, and gets lower precdence in the dispatcher ordering.
```

Comment:
I'd really love to think about how we can simplify this. One way to do it is make it so that FakeTensor isn't actually a subclass; it's some new concept of a non-subclass way to override the behavior of a Tensor (prefer composition over inheritance!) Another is to make it so that FakeTensor doesn't actually override `__torch_dispatch__`, instead, it gets some special magic method `__torch_fake_dispatch__` which slots into the fake tensor spot. Then FakeTensor isn't eligible for dispatch anymore, and you just know to call into this magic method after you've finished subclass resolution  and mode resolution. BTW, we need a giant diagram showing what order things execute after this PR..

Reply:
My first suggestion is quite a lot of work, but I like the second suggestion, it seems conceptually coherent with the rest of what's going on.

Reply:
Agreed I dislike this entire section of code and would love to simplify (and agreed that I'm on the hook for a giant diagram some time after this PR lands 😛).

The second suggestion definitely sounds cleaner than what I have today - I'll think some more about a reasonable way to implement it.

One downside: it relaxes the constraint here that every `at::Tensor with the `Python` key in its key_set must have a `__torch_dispatch__` method defined: https://github.com/pytorch/pytorch/blob/273ad1dd23b2637eeb0fc61881241d0814c9dcce/torch/csrc/utils/python_arg_parser.cpp#L319

Also: Need to think about how this should work for `TorchProxyDispatchMode` and `FunctionalTensor` + `FunctionalTensorMode`. Should they all get... a `__torch_proxy_dispatch__` and `__torch_functional_dispatch__` method? It's not immediately clear to me that this removes more complexity than it adds.